src/layout/umap.c

```diff
@@ -0,0 +1,1038 @@
+            from = IGRAPH_FROM(graph, eid);
+            to = IGRAPH_TO(graph, eid);
+        } else {
+            to = IGRAPH_FROM(graph, eid);
+            from = IGRAPH_TO(graph, eid);
+        }
+
+        /* Current coordinates of both vertices */
+        for (int d = 0; d != ndim; d++) {
+            VECTOR(from_emb)[d] = MATRIX(*layout, from, d);
+            VECTOR(to_emb)[d] = MATRIX(*layout, to, d);
+            VECTOR(delta)[d] = MATRIX(*layout, from, d) - MATRIX(*layout, to, d);
+        }
+
+        /* Apply attractive force since they are neighbors */
+        IGRAPH_CHECK(igraph_i_umap_attract(&delta, VECTOR(*umap_weights)[eid], a, b, &forces));
+        for (int d = 0; d != ndim; d++) {
+            MATRIX(*layout, from, d) += learning_rate * VECTOR(forces)[d];
+        }
+
+        /* Random other nodes are repelled from one (the first) vertex */
+        IGRAPH_CHECK(igraph_random_sample(&negative_vertices, 0, no_of_nodes - 2, n_random_vertices));
+        for (igraph_integer_t j = 0; j < n_random_vertices; j++) {
+            /* Get random neighbor */
+            to = VECTOR(negative_vertices)[j];
+            /* obviously you cannot repel yourself */
+            if (to >= from) {
+                to++;
+            }
+            /* do not repel neighbors for small graphs, for big graphs this
+             * does not matter as long as the k in knn << number of vertices */
+            if (avoid_neighbor_repulsion) {
+                /* FIXME: This algo terribly inefficient, try to improve using
+                 * adjacency lists or something but just for testing anyway */
+                igraph_bool_t skip = 0;
+                igraph_incident(graph, &neis, from, IGRAPH_ALL);
+                for (igraph_integer_t k = 0; k < igraph_vector_int_size(&neis); k++) {
+                    igraph_integer_t eid2 = VECTOR(neis)[k];
+                    igraph_integer_t from2, to2;
+                    from2 = IGRAPH_FROM(graph, eid2);
+                    to2 = IGRAPH_TO(graph, eid2);
+                    if (((from2 == from) && (to2 == to)) || ((from2 == to) && (from == to2))) {
+                        skip = 1;
+                        break;
+                    }
+                }
+                if (skip == 1) {
+                    continue;
+                }
+            }
+
+            /* Get layout of random neighbor and gradient in embedding */
+            for (int d = 0; d != ndim; d++) {
+                VECTOR(to_emb)[d] = MATRIX(*layout, to, d);
+                VECTOR(delta)[d] = MATRIX(*layout, from, d) - MATRIX(*layout, to, d);
+            }
+
+            /* This repels the other vertex assuming it's a negative example
+             * that is no weight, no edge */
+            IGRAPH_CHECK(igraph_i_umap_repel(&delta, 0, a, b, &forces));
+            /* The repulsive force is already *away* from the other (non-neighbor) vertex */
+            for (int d = 0; d != ndim; d++) {
```

Comment:
Do not use `int`. I probably didn't catch all uses, please replace them all, except where `int` is truly needed (e.g. LAPACK error code).